OTEP: Logger.Enabled (#4290)
open-telemetry/opentelemetry-specification: oteps/logs/4290-logger-enabled.md

   when emitting a log or event record is unnecessary.
3. Configure a minimum a log severity level on the SDK level.
4. Filter out log and event records when they are inside a span
   that has been sampled out (span is valid and has sampled flag of `false`).
5. **Efficiently** support high-performance logging destination
   like [Linux user_events](https://docs.kernel.org/trace/user_events.html)
   and [ETW (Event Tracing for Windows)](https://learn.microsoft.com/windows/win32/etw/about-event-tracing).
6. Allow **fine-grained** filtering control for logging pipelines
   when using an OpenTelemetry Collector is not feasible
   e.g., for mobile devices, serverless, or IoT.

Without a `Logger.Enabled` check in the OpenTelemetry Logs API
and corresponding implementations in the SDK,
achieving this goal is not feasible.

Address [Specify how Logs SDK implements Enabled #4207](https://github.com/open-telemetry/opentelemetry-specification/issues/4207).

## Explanation

For (1) (2), the user can use the Logs API `Logger.Enabled` function,
which tells the user whether a `Logger` for given arguments
is going to emit a log record.

For (3), the user can declaratively configure the Logs SDK

> using `LoggerConfigurator` to set the `minimum_severity_level`

Author: > We'll need to think how to structure them on LoggerConfig API so it's extendable but only introduce options we have an immediate need for).

Adding a field is a backwards compatible change. I agree that we should add new fields (options) only if we have a big demand for it.

> we'd need to consider having a callback in addition to pure declarative config

For more ad I think the `LogRecordProcessor` comes into play. It should handle the most advanced scenarios.

I will do my best to describe it "Future possibilities" section.

> using `LoggerConfigurator` to set the `minimum_severity_level`

Author: > Also we'd need to consider having a callback in addition to pure declarative config

I added "Alternatives: Dynamic Evaluation in LoggerConfig"

> I think we'll (eventually) have more config options for severity, sampling, and other things, e.g. a few extra scenarios I can think of:

I added "Future possibilities: Extending LoggerConfig

> using `LoggerConfigurator` to set the `minimum_severity_level`

Author: @lmolkova, is there anything more to address regarding this discussion?

   like [Linux user_events](https://docs.kernel.org/trace/user_events.html)
   and [ETW (Event Tracing for Windows)](https://learn.microsoft.com/windows/win32/etw/about-event-tracing).
6. Allow **fine-grained** filtering control for logging pipelines
   when using an OpenTelemetry Collector is not feasible
   e.g., for mobile devices, serverless, or IoT.

Without a `Logger.Enabled` check in the OpenTelemetry Logs API
and corresponding implementations in the SDK,
achieving this goal is not feasible.

Address [Specify how Logs SDK implements Enabled #4207](https://github.com/open-telemetry/opentelemetry-specification/issues/4207).

## Explanation

For (1) (2), the user can use the Logs API `Logger.Enabled` function,
which tells the user whether a `Logger` for given arguments
is going to emit a log record.

For (3), the user can declaratively configure the Logs SDK
using `LoggerConfigurator` to set the `minimum_severity_level`
of a `LoggerConfig`.

For (4), the user can use the Tracing API to check whether
there is a sampled span in the current context before creating
and emitting a log record.
However, the user can may also want to declaratively configure the Logs SDK

> using `LoggerConfigurator` to set the `disabled_on_sampled_out_spans`

Author: I like this name.

> using `LoggerConfigurator` to set the `disabled_on_sampled_out_spans`

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4290/commits/216a21875453b0438dc639e1a03bbfb9564c84b8

within sampled spans.

The addition of `LogRecordProcessor.Enabled` is necessary for
use cases where filtering is dynamic and coupled to processing,
such as (5) and (6).

Both `LoggerConfig` and registered `LogRecordProcessors` take part
in the evalaution of the `Logger.Enabled` operation.
If it returns `true`, meaning the logger is enabled,
then the `LogRecordProcessors` are evaluated.
In such cases, `false` is returned only if all registered
`LogRecordProcessors` return `false` in their `Enabled` calls,
as this means no processor will process the log record.
Pseudo-code:

<!-- markdownlint-disable no-hard-tabs -->
```go
func (l *logger) Enabled(ctx context.Context, param EnabledParameters) bool {
	config := l.config()
	if config.Disabled {
		// The logger is disabled.
		return false
	}
	if params.Severity > config.MinSeverityLevel {
		// The severity is less severe than the logger minimum level.
		return false
	}
	if config.DisabledNotSampledSpans && !trace.SpanContextFromContext(ctx).IsSampled() {
		// The logger is disabled on sampled out spans.
		return false
	}

	processors := l.provider.processors()
	for _, p := range processors {

> 		if p.Enabled(ctx, param) {

Reviewer: what would you expect processors to do inside `Enabled`? If this is a delegating processor, would it call the underlying one? 

I.e. it sounds like there should be just one processor in the pipeline that decides if log should be created. If so, should it be a processor responsibility at all or can we call this component `LogFilter`, `LogSampler`, etc and make it a configurable behavior independent of the processor?

If this component returns true, all processing pipelines would get log record and can drop it if they want to filter in addition. 
If it returns false, none of them should. This is the same as you have outlined in the otep, just extracts filtering from processing.

Reviewer: Some examples of logging filters: 

Here's a good old [.NET ILogger filter](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-9.0#filter-function)

```csharp
builder.Logging.AddFilter((provider, category, logLevel) =>
{
    if (provider.Contains("ConsoleLoggerProvider")
        && category.Contains("Controller")
        && logLevel >= LogLevel.Information)
    {
        return true;
    }
    else if (provider.Contains("ConsoleLoggerProvider")
        && category.Contains("Microsoft")
        && logLevel >= LogLevel.Information)
    {
        return true;
    }
    else
    {
        return false;
    }
});
```

or good old [log4j](https://logback.qos.ch/manual/filters.html)

```java
private void updateLoggers(final Appender appender, final Configuration config) {
    final Level level = null;
    final Filter filter = null;
    for (final LoggerConfig loggerConfig : config.getLoggers().values()) {
        loggerConfig.addAppender(appender, level, filter);
    }
    config.getRootLogger().addAppender(appender, level, filter);
}
```

I.e. composition instead of inheritance - when the logging is configured, each appender/provider pipeline gets dedicated filter independent of the appender/provider logic.

Reviewer: > For (5) would someone configuring an user_events exporter would have to configure a filterer as well as processor?

The trade-off: 
1. most processors would not need `IsEnabled` - processors today are for processing LogRecords. If `IsEnabled` is an API on the processor, from now on everyone would need to think twice whether they need to implement it and what would delegating/composite processors need to do (any of or all of?)
2. a few processors (really exporters, not processors) that would need `IsEnabled` would have a bit more complicated config 

I would pick 2 as the better option.

> For (6) it would not be even possible to e.g. allowing distinct minimum severity levels for different processorsas the filters would run before the processors.

The filter can be "any of" (if any of the filters returns true, record is passed to the processor)  or Processor implementations can absolutely have additional filtering, but not through base interface API.

Reviewer 2: Can processors that need additional functionality implement both interfaces?  Or is that not possible in all languages?

Author: > Can processors that need additional functionality implement both interfaces? Or is that not possible in all languages?

Do you mean something that is currently described in "Trade-offs and mitigations" section?

> For some langagues extending the `LogRecordProcessor` may be seen as breaking.
> For these languages implementing `LogRecordProcessor.Enabled` must be optional.
> The SDK `LogRecordProcessor` must return `true` if `Enabled` is not implemented.
> This approach is currently taken by OpenTelemetry Go.

Is it what you have in mind?
If so then do you think it should be the recommended approach? 
I would say that it is a language/SIG specific decision because of the trade-offs of either approaches.

On the other hand, based on the feedback, I think that we should indeed try to coin define a new filterer abstraction (interface) that could be used as an option for the logger provider and processors. Defining it as a separate abstraction may result in better user experience in most scenarios.

My plan is to revisit, refresh and enhance a prototype I did some time ago: https://github.com/open-telemetry/opentelemetry-go/pull/5825

Author: After some thoughts and chat with @lmolkova, I think we can start with `LogRecordProcessor.Enabled`. 

I still plan to create an issue for "Separate LogRecordFilterer Abstraction". If we will get feedback that `LogRecordProcessor.Enabled` is bad then we can switch to it. I can still make a prototype for the alternative design. We can always revisit the design during Development phase before going stable after gathering feedback and more hand-on experience.

Author: I did my best to address it via https://github.com/open-telemetry/opentelemetry-specification/pull/4290/commits/1a977c6756d8936587e1ffe28a71f7d6d6d78d6c

6. Allow **fine-grained** filtering control for logging pipelines
   when using an OpenTelemetry Collector is not feasible
   e.g., for mobile devices, serverless, or IoT.

Without a `Logger.Enabled` check in the OpenTelemetry Logs API
and corresponding implementations in the SDK,
achieving this goal is not feasible.

OpenTelemetry Logs SDK users require:

- A declarative configuration to conveniently address
  the most popular use cases.
- A dynamic hook for advanced customization and flexibility.

Address [Specify how Logs SDK implements Enabled #4207](https://github.com/open-telemetry/opentelemetry-specification/issues/4207).

## Explanation

For (1) (2), the user can use the Logs API `Logger.Enabled` function,
which tells the user whether a `Logger` for given arguments
is going to emit a log record.

For (3), the user can declaratively configure the Logs SDK
using `LoggerConfigurator` to set the `minimum_severity_level`
of a `LoggerConfig`.

For (4), the user can use the Tracing API to check whether
there is a sampled span in the current context before creating
and emitting a log record.
However, the user can may also want to declaratively configure the Logs SDK

> using `LoggerConfigurator` to set the `trace_based`

Reviewer: I agree with the intent.

`LoggerConfigurator.trace_based` seems very narrow. I still think we need more than a boolean flag.  

I'd model it in the following way (we can take one composite filter vs filter per processor discussion aside)

```yaml
logger_provider:
  filters: # it should be possible to create composite filter on the API, no hard opinion on the config
    trace_based:
      # similarly to span samplers, it can have addition config
      root: # what to do if there is no current span?
         always_on | always_off
    level_based: # I really think this should be top-level property, just entertaining the idea
      min_level: INFO
    rate_based: # throttling
       max_rate_per_scope: ... 
    context_based: # something else that may come up in the future
```

I'm not saying we need to describe all of it, but I do think we need to model API and config for trace-based thing in the extendable way. 

Wonder if @open-telemetry/configuration-approvers would have any advice on how to model the config part of it (and whether it makes sense to model it similarly to [samplers](https://github.com/open-telemetry/opentelemetry-configuration/blob/bbe4bf1ea1aa2282f0f63d4182111ffb74d2d481/examples/sdk-config.yaml#L105))

Author: I will change the OTEP so that `LoggerConfig..trace_based` lands in "Future possibilities".

Author: Addressed here 80ef784

and to extend the `LogRecordProcessor` with an `Enabled` operation.

The addition of `LoggerConfig.minimum_severity_level` is supposed
to serve the (3) use case in an easy-to-setup and efficient way.

The addition of `LoggerConfig.trace_based` can serve the (4)
use case in a declarative way configured on the SDK level
if the user would want to only capture the log records that are
within sampled spans.

The addition of `LogRecordProcessor.Enabled` is necessary for
use cases where filtering is dynamic and coupled to processing,
such as (5) and (6).

Both `LoggerConfig` and registered `LogRecordProcessors` take part
in the evalaution of the `Logger.Enabled` operation.
If it returns `true`, meaning the logger is enabled,
then the `LogRecordProcessors` are evaluated.
In such cases, `false` is returned only if all registered
`LogRecordProcessors` return `false` in their `Enabled` calls,
as this means no processor will process the log record.
Pseudo-code:

<!-- markdownlint-disable no-hard-tabs -->
```go
func (l *logger) Enabled(ctx context.Context, param EnabledParameters) bool {
	config := l.config()
	if config.Disabled {
		// The logger is disabled.
		return false
	}

> 	if params.Severity > config.MinSeverityLevel {

Reviewer 2: If logger.Enabled is fully pluggable anyways, should min severity and trace_based just be processors (or filters)?

Author: With  [`LoggerConfigurator`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#loggerconfigurator) you are not able to change/apply processors.

Technically, there is nothing preventing having both `LoggerConfig.minimum_severity_level` and `MinimumSeverityLevelProcessor` (or `MinimumSeverityLevelFilterer)`. One is a configuration for loggers and the second for processing.
`LoggerConfig.minimum_severity_level` would be a good choice when one would like to change the minimum servierty level for loggers named `MyCompany.NoisyModule.*`.
`MinimumSeverityLevelProcessor` (or `MinimumSeverityLevelFilterer)` would be the **only** choice when one would like to set the minimum servierty level for a certain exporting pipeline (e.g. on batching processor would be exporting .

Let me know what you think. If you think it makes sense then I will describe it in the OTEP.

Author: @dashpole, technically, there is nothing preventing having both `LoggerConfig.minimum_severity_level` and `MinimumSeverityLevelProcessor` (or `MinimumSeverityLevelFilterer)`. One is a configuration for loggers and the second for log processing pipelines.

`LoggerConfig.minimum_severity_level` would be a good choice when one would like to change the minimum severity level for loggers named `MyCompany.NoisyModule.*`. With  [`LoggerConfigurator`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#loggerconfigurator) you are not able to change/apply processors so we cannot use the `LoggerConfig` to set processors.

`MinimumSeverityLevelProcessor` (or `MinimumSeverityLevelFilterer)` would be the **only** choice when one would like to set the minimum severity level for a certain exporting pipeline (e.g. on batching processor would be exporting .

Let me know what you think. If you think it makes sense then I will describe it in the OTEP.

CC @jack-berg

Author: I tried to describe it in the OTEP e96f9e7e4c10c84b88cc460df10ca817d16961e3

@dashpole, does it address your feedback?

Reviewer 2: It does, thanks.  A few more thoughts. Take them with a grain of salt as I haven't thought through everything:

* It (obviously) is a bit be confusing to have both MinimumSeverityLevelFilterer and LoggerConfig.minimum_severity_level.
* It would be nice to be able to apply arbitrary filters to a specific Logger instead of having a list of predefined filters in LoggerConfig.  The spec can always define a list of required filters that yaml config can rely on.
    * Note: Processing doesn't make sense to me at a logger level.
* I know it is a common pattern today for loggers to have their minimum_severity_level increased before being passed to a library to disable spammy logs. But I do like how `View`s, provide access to scope information for config at the MeterProvider level, rather than providing config at the Meter level. I wonder if we've considered that as a way to simplify the config surface (and not have config at both Logger and LoggerProvider levels).

Reviewer 2: Ah, sounds like this discussion is mostly orthogonal to this OTEP.

> since the purpose of `LoggerConfig` is static configuration

I don't quite understand this. Is there a performance difference between the two? Or is the idea that we don't want to mix static config (e.g. disabled) with functional config?

Author: > I don't quite understand this. 

I should remove, rephrase it.

Author: @dashpole, PTAL 48c1e110fec85e2ca1793117141e960d40a52e36

Reviewer 2: sgtm

# Logger.Enabled

## Motivation

In applications requiring high performance,
while optimizing the performance of enabled logging is important,
it is often equally or more critical to minimize the overhead of logging
for disabled or un-exported logs.

The consumers of OpenTelemetry clients want to:

1. **Correctly** and **efficiently** bridge features
   like `LogLevelEnabled` in log bridge/appender implementations.
2. Avoid allocating memory to store a log record,
   avoid performing computationally expensive operations,
   and avoid exporting
   when emitting a log or event record is unnecessary.

> 3. Configure a minimum a log severity level on the SDK level.

Reviewer: ```suggestion
3. Configure a minimum log severity level on the SDK level.
```

For (1) (2), the user can use the Logs API `Logger.Enabled` function,
which tells the user whether a `Logger` for given arguments
is going to emit a log record.

For (3), the user can declaratively configure the Logs SDK
using `LoggerConfigurator` to set the `minimum_severity_level`
of a `LoggerConfig` for given `Logger`.

For (4), the user can use the Tracing API to check whether
there is a sampled span in the current context before creating
and emitting a log record.

For (5) (6), the user can hook to `Logger.Enabled` Logs API calls
by adding to the Logs SDK a `LogRecordProcessor` implementing `Enabled`
(or alternatively other hook such as `LogRecordFilterer`;
see [here](#separate-logrecordfilterer-abstraction)).

## Internal details

Regarding (1) (2), the Logs API specification has already introduced `Logger.Enabled`:

- [Add Enabled method to Logger #4020](https://github.com/open-telemetry/opentelemetry-specification/pull/4020)
- [Define Enabled parameters for Logger #4203](https://github.com/open-telemetry/opentelemetry-specification/pull/4203)

The addition of `LoggerConfig.minimum_severity_level` is supposed
to serve the (3) use case in an easy-to-setup and efficient way.

> Regarding (4) the API callers can decide whether to emit log records
> for spans that are not sampled. Example in Go:

Reviewer: I think it's ok for demonstration purposes, but general-purpose instrumentations should not hardcode log-if-span-is-sampled behavior. It should be application choice.

maybe 
```suggestion
Regarding (4) the application developers can decide whether to emit log records
for spans that are not sampled. Example in Go:
```

?

<!-- markdownlint-disable no-hard-tabs -->
```go
if trace.SpanContextFromContext(ctx).IsSampled() && logger.Enabled(ctx, params) {
	logger.Emit(ctx, createLogRecord(payload))
}
```
<!-- markdownlint-enable no-hard-tabs -->

For instrumentation libraries, the API-level control might be more appropriate,

> than configuring such behavior on the SDK level.

Reviewer: I'm not sure what this sentence means, could you please elaborate?

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4290/files#r1916246180

I will elaborate.

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4290/commits/76d59e3f9ff7347de92daa7ffb68925a58b47971

I think this would be better.

Reviewer: > The instrumentation libraries can have a configuration defining
if they emit log records that are not inside sampled spans.

I'd prefer to remove this sentence - imagine how painful it would be to configure if every instrumentation exposed a setting like this. We should not ask instrumentation libs to do it - it should only be done on the SDK level (we just don't specify how yet)

Author: Removed https://github.com/open-telemetry/opentelemetry-specification/pull/4290/commits/bb36dd38f4da9bf00fa8394eb0000b05db9ad777

The addition of `LogRecordProcessor.Enabled` is necessary for
use cases where filtering is dynamic and coupled to processing,
such as (5) and (6).

Both `LoggerConfig` and registered `LogRecordProcessors` take part
in the evalaution of the `Logger.Enabled` operation.

> If it returns `true`, meaning the logger is enabled,

Reviewer: nit: 

```suggestion
If `LoggerConfig` returns `true`, meaning the logger is enabled,
```

is this correct?

Author: ```suggestion
If `LoggerConfig.disabled` is `false`, meaning the logger is enabled,
```

# Logger.Enabled

## Motivation

In applications requiring high performance,
while optimizing the performance of enabled logging is important,
it is often equally or more critical to minimize the overhead of logging
for disabled or un-exported logs.

The consumers of OpenTelemetry clients want to:

1. **Correctly** and **efficiently** bridge features
   like `LogLevelEnabled` in log bridge/appender implementations.
2. Avoid allocating memory to store a log record,
   avoid performing computationally expensive operations,
   and avoid exporting
   when emitting a log or event record is unnecessary.

> 3. Configure a minimum log severity level on the SDK level.

Reviewer 3: Ideally, this is configurable on a per-processor level, not just SDK wide, so you can have alternate destinations for different log severity.

Author: I agree. I will add this use case. Also given later I already described:

> `MinimumSeverityLevelProcessor` is for configuring log processing pipelines.
It is the only choice when one would like to set the minimum severity level
for a certain exporting pipeline.
For example, one batching processor would be exporting logs only above info level
via OTLP and a second simple processor would be exporting all logs to stdout.

> 3. Configure a minimum log severity level on the SDK level.

Reviewer 4: > 3. Configure a minimum log severity level on the SDK level.

Sharing my experience from OTel .NET, Rust. Such a capability is already provided by the existing Logging library.
For example, the following one-liner is enough to tell .NET's ILogger to only send INFO and above to OpenTelemetry.
```csharp
builder.AddFilter<OpenTelemetryLoggerProvider>("*", LogLevel.INFO);
```

Similar in Rust. The following tells the logging library to send only Info and above to OTel
```rust
let filter_otel = EnvFilter::new("info");
let otel_layer = otel_layer.with_filter(filter_otel);
```

Can you share scenarios where this functionality had to be provided by OTel itself? Is this only for scenarios where users use OTel Logs/Event API and they need to achieve filtering?

Author: > Is this only for scenarios where users use OTel Logs/Event API and they need to achieve filtering?
 
This is one of the reasons. Thanks to it we minimize the overhead caused by bridge. Making complex OTel Events could cause significant overhead because it's the conversions.

For OTel Semantical events and log records pecific processing/filtering it would be easier for the users to use the OTel Data Model.

We could provide filtering and processors common that are related to OTel domain. E.g. filtering by span context. If we do not have this in the SDK then we would need to create such filter for each library.

Finally, I guess we would like to have some support via OTel Configuration files in future.

4. Filter out log and event records when they are inside a span
   that has been sampled out (span is valid and has sampled flag of `false`).
5. **Efficiently** support high-performance logging destination
   like [Linux user_events](https://docs.kernel.org/trace/user_events.html)
   and [ETW (Event Tracing for Windows)](https://learn.microsoft.com/windows/win32/etw/about-event-tracing).
6. Allow **fine-grained** filtering control for logging pipelines
   when using an OpenTelemetry Collector is not feasible
   e.g., for mobile devices, serverless, or IoT.

Without a `Logger.Enabled` check in the OpenTelemetry Logs API
and corresponding implementations in the SDK,
achieving this goal is not feasible.

OpenTelemetry Logs SDK users require:

- A declarative configuration to conveniently address

>   the most popular use cases.

Reviewer 3: This seems to be doing too much. Why not just focus on adding logger enabled like the title indicates?

- A dynamic hook for advanced customization and flexibility.

The main purpose of this OTEP is to have foundations for:

- [#4364](https://github.com/open-telemetry/opentelemetry-specification/issues/4364):
  extending the SDK's `LoggerConfig` with `min_severity` field,
  
- [#4363](https://github.com/open-telemetry/opentelemetry-specification/issues/4363):
  extending the SDK's `LogRecordProcessor` with an opt-in `Enabled` operation,

and address [Specify how Logs SDK implements Enabled #4207](https://github.com/open-telemetry/opentelemetry-specification/issues/4207).

## Explanation

> For (1) (2), the user can use the Logs API `Logger.Enabled` function,

Reviewer 3: These number references are confusing as the seem to be referencing something from another section.

which tells the user whether a `Logger` for given arguments
is going to emit a log record.

For (3), the user can declaratively configure the Logs SDK
using `LoggerConfigurator` to set the `min_severity`
of a `LoggerConfig` for given `Logger`.

For (4), the user can use the Tracing API to check whether

> there is a sampled span in the current context before creating

Reviewer 4: is the user here referring to end-users? Are you expecting users to do this pattern? 

do_something_important()
{
  logger.log("I did something important", results = {results()});
}

do_something_important()
{
  if(Span::GetActiveSpan().IsRecording())
  {
       logger.log("I did something important", results = {results()});
  }
  else
 {
  //don't do logging.
 }
}

I think this should be a SDK setting and instrumentations should not have to worry about this. They always call logger.log. Upto SDK to decide (based on config) if the log must be dropped as the active span is not recorded.

(Sorry If I misunderstood the section!)

Author: Yes. It could be also implemented via a Processor which filters out not recorded spans.

and emitting a log record.

For (5) (6), the user can hook to `Logger.Enabled` Logs API calls
by adding to the Logs SDK a `LogRecordProcessor` implementing `Enabled`
(or alternatively other hook such as `LogRecordFilterer`;
see [here](#separate-logrecordfilterer-abstraction)).

## Internal details

Regarding (1) (2), the Logs API specification has already introduced `Logger.Enabled`:

- [Add Enabled method to Logger #4020](https://github.com/open-telemetry/opentelemetry-specification/pull/4020)
- [Define Enabled parameters for Logger #4203](https://github.com/open-telemetry/opentelemetry-specification/pull/4203)

The addition of `LoggerConfig.min_severity` is supposed
to serve the (3) use case in an easy-to-setup and efficient way.

Regarding (4) the application developers can decide whether to emit log records
for spans that are not sampled. Example in Go:

<!-- markdownlint-disable no-hard-tabs -->
```go
if trace.SpanContextFromContext(ctx).IsSampled() && logger.Enabled(ctx, params) {
	logger.Emit(ctx, createLogRecord(payload))
}
```
<!-- markdownlint-enable no-hard-tabs -->

The addition of `LogRecordProcessor.Enabled` is necessary for

> use cases where filtering is dynamic and coupled to processing,

Reviewer 4: I'd like to get more clarification on this.

For 5 (when exporting to OS kernel facilities like etw/user-events), this is primarily a perf saving mechanism. If OS says nobody is interested in your log/event, then the producer can short circuit. This has to be done very quickly and typically only has 1-2 top-level parameter like Severity/EventName/ScopeName. 

6 is referring to scenarios where users want to do filtering, primarily to control volume of telemetry exported. This usually is done based on LogRecord (i.e body/attributes/name/severity etc.). I believe this is already doable by writing normal LogRecordProcessors, like the one shown in https://github.com/open-telemetry/opentelemetry-specification/pull/4407

In other words, I see 5 requires LogProcessor.Enabled, but 6 does not require.

Author: Thanks. My description of 6 is indeed not clear. I meant that the perf saving mechanism works when the procesors are composed - supports the advanced scenarios which I described in https://github.com/open-telemetry/opentelemetry-specification/pull/4407

	for _, f := range fltrs {
		if f.Enabled(ctx, param) {
			// At least one Processor will process the Record.
			return true
		}
	}
	// No Processor will process the record
	return false
}
```
<!-- markdownlint-enable no-hard-tabs -->

The implementation of `LogRecordProcessor.Enabled` is optional
so that we can optimize (and not loop though processors)
when the processor does not implement it.

There is nothing preventing having both `LoggerConfig.min_severity`
and something like a `MinimumSeverityLevelProcessor`.

`LoggerConfig.min_severity` is a configuration for concrete loggers.
For instance, it would be a easy to use feature when one would like to change
the minimum severity level for all loggers with names matching
`MyCompany.NoisyModule.*` wildcard pattern.
With `LoggerConfigurator` the user is not able to change/apply a processor.

`MinimumSeverityLevelProcessor` is for configuring log processing pipelines.
It is the only choice when one would like to set the minimum severity level
for a certain exporting pipeline.
For example, one batching processor would be exporting logs only above info level

> via OTLP and a second simple processor would be exporting all logs to stdout.

Author: Having two "layers" of setting up the severity filtering may be very confusing for the users. 
The behavior of `LoggerConfig.min_severity` could be achieved also via a filter processor that takes into consideration the `InstrumentationScope` when filtering by severity level.

Most probably for languages like Java `LoggerConfig.min_severity` won't be helpful as the severity level is already set on the bridged logging libraries.

The more important piece is to support cases like described here: https://github.com/open-telemetry/opentelemetry-specification/pull/4290/files#r1925688450

@jack-berg, @trask, any opinion?

@jack-berg, are all OTel SDKs are supposed support `LoggerConfig` (assuming it gets stabilized) or is it is assumed as an optional feature?

CC @MrAlias

Author: Side note: I think that `LoggerConfig` functionality could be replaced with a `LogRecordProcessor` decorator that implements the same thing. I may try to prototype it.

Author: > I think that LoggerConfig functionality could be replaced with a LogRecordProcessor decorator that implements the same thing. I may try to prototype it.

Here is a PR for OTel Go that improves the filtering processor so that it has instrumentation scope and resource in its input. 

- https://github.com/open-telemetry/opentelemetry-go/pull/6286/

Therefore, it can replace the the need of `LoggerConfig.min_severity` as one can make a processor that just filters the log records by severity and instrumentation scope.

# Logger.Enabled

> ## Motivation

Reviewer 4: I'd suggest adding something to the tune of:
"In applications requiring extremely high performance, it is often more important to ensure the cost of logging is zero/close-to-zero when they are not enabled, than optimizing the cost of logging when that log is of interest. Without an IsEnabled check at the entry point on OTel LogBridge API and corresponding implementations in SDK, it is not possible to meet this goal"

Author: I will work on that for sure. Thanks for giving me the idea.

The consumers of OpenTelemetry clients want to:

1. Correctly and efficiently bridge features like `LogLevelEnabled` in log bridge/appender implementations.

> 2. Avoid performing computationally expensive operations and exporting when emitting a log or event record is unencessary.

Reviewer 4: please add: avoid allocating memory unnecessarily to store LogRecord

Author: WDYT about the following?

```suggestion
2. Avoid allocating memory unnecessarily to store a log record, avoid performing computationally expensive operations and avoid exporting when emitting a log or event record is unnecessary.
```

Reviewer 4: good!

> 3. Control a minmium a log serverity level on the SDK level.

Reviewer 4: ```suggestion
3. Control a minimum a log serverity level on the SDK level.
```

> 4. Filter out log and event records when they are not inside a recording span.

Reviewer 4: ```suggestion
4. Filter out log records when they are not inside a recording span.
```

Should be good enough as SDK has no separate event now?

Author: I wanted to explicitly mention that it can be for all possible cases.

5. Have fine-grained filtering control for logging pipelines without using an OpenTelemetry Collector (e.g. mobile devices, serverless, IoT).

> 6. Efficiently support high-performance logging destionation like Linux user_events and ETW (Event Tracing for Windows).

Reviewer 4: ```suggestion
6. Efficiently support high-performance logging destination like Linux user_events and ETW (Event Tracing for Windows).
```

7. Add sampling for logging.

## Explanation

For (1) (2), the user can use the Logs API `Logger.Enabled` function, which tells the user wheteher a `Logger` for given arguments is going to emit a log record.

For (3) (4), the user can declarativly configure the Logs SDK using `LoggerConfigurator` to set the `disabled`, `minimum_severity_level`, `disabled_not_recorded_spans` of a `LoggerConfig`.

> For (5) (6) (7), the user can hook to `Logger.Enabled` Logs API calls by adding to the Logs SDK a `LogRecordProcessor` implementing `OnEnabled`.

Reviewer 4: I'd add `LogRecordExporter` implementing `OnEnabled` too, though technically a processor can do the job

Author: I do not think this makes sense for "batching" exporters. I see `LogRecordExporter` as an abstraction that is suited for batching.

Author: I added https://github.com/open-telemetry/opentelemetry-specification/pull/4290/commits/bdc1eea6dd0470f8d221e3611a80034ccc925c75

> For (5) (6) (7), the user can hook to `Logger.Enabled` Logs API calls by adding to the Logs SDK a `LogRecordProcessor` implementing `OnEnabled`.

Author: I think most of the points are not obvious and would need more detailed description. I was mostly focused on listing the motivation and my vision of the "design goal" before going too much into details.

> For (5) (6) (7), the user can hook to `Logger.Enabled` Logs API calls by adding to the Logs SDK a `LogRecordProcessor` implementing `OnEnabled`.

Reviewer 4: Got it. The efficiency part is the important one, so perhaps make it **bold**

# Logger.Enabled

## Motivation

In applications requiring extremely high performance,
minimizing the cost of logging when it is not exported is often more critical

> than optimizing the cost of logging when the log record is of interest.

Reviewer 4: In applications requiring high performance, while optimizing the performance of enabled logging is important, it is often equally or more critical to minimize the overhead of logging for disabled or un-exported logs.

> In applications requiring extremely high performance,
> minimizing the cost of logging when it is not exported is often more critical
> than optimizing the cost of logging when the log record is of interest.

Author: ```suggestion
In applications requiring high performance,
while optimizing the performance of enabled logging is important,
it is often equally or more critical to minimize the overhead of logging
for disabled or un-exported logs.
```

The consumers of OpenTelemetry clients want to:

1. **Correctly** and **efficiently** bridge features like `LogLevelEnabled` in log bridge/appender implementations.
2. Avoid allocating memory to store a log record, avoid performing computationally expensive operations and avoid exporting when emitting a log or event record is unnecessary.

> 3. Configure a minimum a log serverity level on the SDK level.

Reviewer 4: ```suggestion
3. Configure a minimum a log severity level on the SDK level.
```

4. Filter out log and event records when they are not inside a recording span.

> 5. Have **fine-grained** filtering control for logging pipelines without using an OpenTelemetry Collector (e.g. mobile devices, serverless, IoT).

Author: I do not expect the Logs API to be as user-friendly as the logging libraries popular for given language.

However, I think that the SDK (the logging backend) should be able handle different logging processing scenarios. E.g. sending verbose logs via user_events and warning logs via OTLP (probably I could come up with better examples). I heard that someone wanted to send non-event log records and event records to different destinations. Moreover, there is the following question in the [Events Basics OTEP](https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/0265-event-vision.md):

> How to support routing logs from the Logs API to a language-specific logging library while simultaneously routing logs from the language-specific logging library to an OpenTelemetry Logging Exporter?

Side note: I think that even without this use case we should aim for achieving a composable design.

> 5. Have **fine-grained** filtering control for logging pipelines without using an OpenTelemetry Collector (e.g. mobile devices, serverless, IoT).

Author: I moved it as the last use-case given that it may be seen as the least important. https://github.com/open-telemetry/opentelemetry-specification/pull/4290/commits/a4257eff7767dc8ce8feebc4b3e9c21a37a570a5

> 5. Have **fine-grained** filtering control for logging pipelines without using an OpenTelemetry Collector (e.g. mobile devices, serverless, IoT).

Author: Let me know if the current way it is described looks good overall.

> 5. Have **fine-grained** filtering control for logging pipelines without using an OpenTelemetry Collector (e.g. mobile devices, serverless, IoT).

Author: @cijothomas, PTAL

I want to also add that some users may want to directly use Logs API without using any bridged library. See: https://github.com/open-telemetry/opentelemetry-specification/pull/4352

> 6. **Efficiently** support high-performance logging destination like Linux user_events and ETW (Event Tracing for Windows).

Reviewer 4: ```suggestion
6. **Efficiently** support high-performance logging destination like [Linux user_events](https://docs.kernel.org/trace/user_events.html) and [ETW (Event Tracing for Windows)](https://learn.microsoft.com/windows/win32/etw/about-event-tracing).
```

7. Add sampling for logging.

Without a `Logger.Enabled` check in the OpenTelemetry Logs API
and corresponding implementations in the SDK,
achieving this goal is not feasible.

## Explanation

> For (1) (2), the user can use the Logs API `Logger.Enabled` function, which tells the user wheteher a `Logger` for given arguments is going to emit a log record.

Reviewer 4: ```suggestion
For (1) (2), the user can use the Logs API `Logger.Enabled` function, which tells the user whether a `Logger` for given arguments is going to emit a log record.
```

1. **Correctly** and **efficiently** bridge features like `LogLevelEnabled` in log bridge/appender implementations.
2. Avoid allocating memory to store a log record, avoid performing computationally expensive operations and avoid exporting when emitting a log or event record is unnecessary.
3. Configure a minimum a log severity level on the SDK level.
4. Filter out log and event records when they are not inside a recording span.
5. Have **fine-grained** filtering control for logging pipelines without using an OpenTelemetry Collector (e.g. mobile devices, serverless, IoT).
6. **Efficiently** support high-performance logging destination like [Linux user_events](https://docs.kernel.org/trace/user_events.html) and [ETW (Event Tracing for Windows)](https://learn.microsoft.com/windows/win32/etw/about-event-tracing).
7. Add sampling for logging.

Without a `Logger.Enabled` check in the OpenTelemetry Logs API
and corresponding implementations in the SDK,
achieving this goal is not feasible.

## Explanation

For (1) (2), the user can use the Logs API `Logger.Enabled` function, which tells the user whether a `Logger` for given arguments is going to emit a log record.

> For (3) (4), the user can declarativly configure the Logs SDK using `LoggerConfigurator` to set the `disabled`, `minimum_severity_level`, `disabled_not_recorded_spans` of a `LoggerConfig`.

Reviewer 5: ```suggestion
For (3) (4), the user can declaratively configure the Logs SDK using `LoggerConfigurator` to set the `disabled`, `minimum_severity_level`, `disabled_on_sampled_out_spans` of a `LoggerConfig`.
```
